### ISSUE TITLE
If I create an index on named window, it may cause the Fire-and-Forget query to fail occasionally

### ISSUE THREAD

**Issue by weizn11** [espertechinc/esper#216]
If I create an index on the named window like this：
```
this.epCompilerArgs.getPath().add(this.epRuntime.getRuntimePath());
        this.epDeployService.deploy(this.epCompiler.compile(
                "@public create window MOAWindowsCreateProcessWindow#time(10 min) as select " +
                        "log_id, " +
                        "unix_timestamp, " +
                        "unix_mtime, " +
                        "computer_name, " +
                        "host_ip, " +
                        "misid, " +

                        "proclogset_type, " +
                        "proclogset_processguid, " +
                        "proclogset_parentguid, " +
                        "proclogset_pid, " +
                        "proclogset_ppid, " +
                        "proclogset_sessionid, " +
                        "proclogset_logonid, " +
                        "proclogset_user, " +
                        "proclogset_processname, " +
                        "proclogset_handlecount, " +
                        "proclogset_path, " +
                        "proclogset_md5, " +
                        "proclogset_sha256, " +
                        "proclogset_parentpath, " +
                        "proclogset_cwd, " +
                        "proclogset_command, " +
                        "proclogset_command_raw, " +
                        "proclogset_fileversion, " +
                        "proclogset_companyname, " +
                        "proclogset_filedescription, " +
                        "proclogset_starttime, " +
                        "proclogset_parent_command, " +
                        "proclogset_parent_command_raw, " +
                        "proclogset_parent_processname, " +
                        "proclogset_parent_md5, " +
                        "proclogset_parent_filedescription, " +
                        "proclogset_parent_user " +

                        "from MOAEdrRawLogs",
                this.epCompilerArgs
                )
        );

        //create index
        this.epCompilerArgs.getPath().add(this.epRuntime.getRuntimePath());
        this.epDeployService.deploy(this.epCompiler.compile(
                "create index MOAWindowsCreateProcessWindowIndex " +
                        "on MOAWindowsCreateProcessWindow(" +

                        "computer_name hash " +
                        ")",
                this.epCompilerArgs
                )
        );
```

Then I called the execute_epl() method to execute a Fire-and-Forget query
```
public static EPCompiled compile_fire_epl(String epl) {
        EPCompiled eplComp;

        compilerMutex.lock();
        try {
            CompilerArguments epCompilerArgs = new CompilerArguments(epConfig);
            epCompilerArgs.getPath().add(EPRuntimeProvider.getDefaultRuntime(epConfig).getRuntimePath());
            eplComp = EPCompilerProvider.getCompiler().compileQuery(epl, epCompilerArgs);
        } catch (EPCompileException e) {
            compilerMutex.unlock();
            System.out.print("[SystemError]-[EsperAnalyticsEngine]-[compile_fire_epl] " +
                    "编译EPL语句失败，异常原因：" +
                    e.getMessage() + "\n" + StringUtils.format_stack_trace(e) + "\n" +
                    "原始EPL语句：[" + epl + "]");
            return null;
        }
        compilerMutex.unlock();

        return eplComp;
    }

    public static EventBean[] execute_epl(String epl) {
        EventBean[] queryEventArray;

        EPCompiled eplComp = EsperAnalyticsEngine.compile_fire_epl(epl);
        if (eplComp == null)
            return null;

        try {
            queryEventArray = EPRuntimeProvider.getDefaultRuntime(epConfig).
                    getFireAndForgetService().executeQuery(eplComp).getArray();
        } catch (Exception e) {
            System.out.print("[SystemError]-[EsperAnalyticsEngine]-[execute_epl] " +
                    "执行EPL语句失败，异常原因：" +
                    e.getMessage() + "\n" + StringUtils.format_stack_trace(e) + "\n" +
                    "原始EPL语句：[" + epl + "]");
            return null;
        }

        return queryEventArray;
    }
```

Occasionally, esper will get an error like this at runtime.
```
[SystemError]-[EsperAnalyticsEngine]-[execute_epl] 执行EPL语句失败，异常原因：null
java.util.ConcurrentModificationException
	at java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:719)
	at java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:742)
	at com.espertech.esper.common.internal.epl.expression.core.ExprNodeUtilityEvaluate.applyFilterExpressionIterable(ExprNodeUtilityEvaluate.java:39)
	at com.espertech.esper.common.internal.epl.fafquery.querymethod.FAFQueryMethodSelectExecUtil.filtered(FAFQueryMethodSelectExecUtil.java:74)
	at com.espertech.esper.common.internal.epl.fafquery.querymethod.FAFQueryMethodSelectExecNoContextNoJoin.execute(FAFQueryMethodSelectExecNoContextNoJoin.java:49)
	at com.espertech.esper.common.internal.epl.fafquery.querymethod.FAFQueryMethodSelect.execute(FAFQueryMethodSelect.java:158)
	at com.espertech.esper.runtime.internal.kernel.service.EPFireAndForgetServiceImpl.executeQueryUnprepared(EPFireAndForgetServiceImpl.java:96)
	at com.espertech.esper.runtime.internal.kernel.service.EPFireAndForgetServiceImpl.executeQuery(EPFireAndForgetServiceImpl.java:42)
	at net.weizn.mtata.storm.esper.engine.EsperAnalyticsEngine.execute_epl(EsperAnalyticsEngine.java:495)
	at net.weizn.mtata.storm.esper.engine.EsperQueryUtils.query_epl(EsperQueryUtils.java:25)
	at net.weizn.mtata.storm.esper.engine.EsperQueryUtils.query_win_below_command_moa(EsperQueryUtils.java:259)
	at net.weizn.mtata.storm.esper.engine.EsperQueryUtils.query_win_context_command_moa(EsperQueryUtils.java:297)
	at net.weizn.mtata.storm.esper.ruleset.MOA_EDR.MOASigmaExtractAlerts$AlertsListener$HandlerThread._event_handler(MOASigmaExtractAlerts.java:224)
	at net.weizn.mtata.storm.esper.ruleset.MOA_EDR.MOASigmaExtractAlerts$AlertsListener$HandlerThread.run(MOASigmaExtractAlerts.java:368)

原始EPL语句：[select * from MOAWindowsCreateProcessWindow where computer_name = 'admin-pc' and proclogset_command_raw is not null and StringUtils.length(StringUtils.strip(proclogset_command_raw, ' ')) > 0 and proclogset_parentguid = '0a09b283-6092-4df0-a208-4ec82354c25b' and proclogset_ppid = 6344 and unix_timestamp >= 1599396347001 and proclogset_pid != 5216 and StringUtils.calc_md5_16(proclogset_parentpath) = '162760b5a5436d45' order by unix_timestamp asc, proclogset_pid asc limit 40 ]
```
But if I delete the statement that created the index, there is no such error. What is the reason for this error?

Thank you very much for your time and help,
Zinanwei

**Comment by bernhardttom** [comment]
Usually there is a context partition lock that covers the index. Its not clear why that wouldn't take. Can you please provide a test class?

**Comment by weizn11** [comment]
> Usually there is a context partition lock that covers the index. Its not clear why that wouldn't take. Can you please provide a test class?

I tried coding some demos in the test environment, but none of them reproduced the problem.
While the code for the production environment is still problematic, I've found that if I change the HASH index to the BTREE index, the error is no longer reported.
Of course,  if I can reproduce the problem in the future, I'll send you the demo.

Thanks very much~
